tools/fifo.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2020 Codethink Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fifo.h"
+
+uint16_t fifo_advance(uint16_t pos)
+{
+	pos++;
+	return (pos >= FIFO_MAX) ? 0 : pos;
+}
+
+uint16_t fifo_seek(int16_t pos, int16_t steps)
+{
+	pos += steps;
+	if (pos >= FIFO_MAX) {
+		pos -= FIFO_MAX;
+	} else if (pos < 0) {
+		pos += FIFO_MAX;
+	}
+	return pos;
+}
+
+// Returns the nth previously added element from the fifo
+bool fifo_peek_back(struct fifo *fifo, uint16_t steps, uint32_t *out)
+{
+	if (fifo->used < steps) {
+		// Can't peek back further than we've written
+		return false;
+	}
+
+	*out = fifo->values[fifo_seek(fifo->read, steps)];
```

Comment:
I may be confused, but should this not be

```suggestion
	*out = fifo->values[fifo_seek(fifo->read, -steps)];
```

?